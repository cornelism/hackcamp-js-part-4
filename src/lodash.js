// TODO: Implement these functions
// Look at what's available on Array and Object to help you.
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array
// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object

/**
 ** This method returns the first argument it receives.
 **/
export const identity = value => value;

/**
 * Iterate over an input object, calling a provided function callback for each key and value in the object.
 **/
export function forEachObjIndexed(object, callback) {
  for (let key in object) {
    callback(object[key], key);
  }
}

/**
 * Creates an array of own enumerable string keyed-value pairs for object
 */
export const entries = object =>
  Object.keys(object).map(key => [key, object[key]]);

/**
 * Creates a new object where every value is mapped using callback
 */
export function map(collection, callback) {
  let array = [];
  collection.forEach((val, index, arr) => {
    array = [...array, callback(val, index, arr)];
  });
  return array;
}

/**
 * This method is like _.find except that it returns the key of the first element predicate returns truthy for instead
 * of the element itself.
 */
export function findKey(object, predicate) {
  if (typeof predicate === 'function') {
    for (let key in object) {
      if (predicate(object[key])) return key;
    }
  } else {
    for (let key in object) {
      if (JSON.stringify(object[key]) === JSON.stringify(predicate)) return key;
    }
  }
}

/**
 * Flattens array a single level deep.
 */
export const flatten = array => array.reduce((acc, val) => acc.concat(val), []);

/**
 * Recursively flattens array.
 */
export const flattenDeep = array =>
  array.reduce(
    (acc, val) =>
      Array.isArray(val) ? [...acc, ...flattenDeep(val)] : [...acc, val],
    []
  );

/**
 * Creates an array with all falsey values removed. The values false, null, 0, "", undefined, and NaN are falsey.
 */
export const compact = array => array.filter(element => !!element);

/**
 * Creates an array of elements split into groups the length of size. If array can't be split evenly, the final chunk
 * will be the remaining elements.
 */
export const chunk = (array, size = 1) =>
  Array.isArray(array)
    ? array.reduce(
        (acc, val, idx) =>
          idx % size === 0 ? [...acc, array.slice(idx, idx + size)] : acc,
        []
      )
    : [];

/**
 * Creates an array of grouped elements, the first of which contains the first elements of the given arrays, the second
 * of which contains the second elements of the given arrays, and so on.
 */
export function zip(...arrays) {
  const filteredArrays = arrays.filter(Array.isArray);
  const length = Math.max(...filteredArrays.map(arr => arr.length));
  return Array.from({length}, (value, index) =>
    filteredArrays.map(
      array => (array.length - 1 >= index ? array[index] : undefined)
    )
  );
}

/**
 * This method accepts two arrays, one of property identifiers and one of corresponding values
 */
export function zipObject(props, values) {
  const obj = {};
  props.map((val, index) => (obj[val] = values[index]));
  return obj;
}

/**
 * Creates an object composed of keys generated from the results of running each element of collection thru iteratee
 */
export function groupBy(collection, iteratee) {
  const obj = {};
  collection.map(item => {
    obj[iteratee(item)] = obj[iteratee(item)]
      ? [...obj[iteratee(item)], item]
      : [item];
  });
  return obj;
}

/**
 * Creates a duplicate-free version of an array
 */
export const uniq = collection =>
  collection.reduce(
    (acc, val) => (acc.includes(val) ? acc : acc.concat(val)),
    []
  );

/**
 * Creates a function that memoizes the result of func. If resolver is provided, it determines the cache key for storing
 * the result based on the arguments provided to the memoized function. By default, the first argument provided to the
 * memoized function is used as the map cache key. The func is invoked with the this binding of the memoized function.
 */
export function memoize(fn) {
  const map = new Map();

  return function(key) {
    if (map.has(key)) {
      return map.get(key);
    }
    map.set(key, fn(...arguments));
    return map.get(key);
  };
}
