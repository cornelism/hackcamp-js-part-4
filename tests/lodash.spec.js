import * as _ from '../src/lodash';

describe('Lodash library', () => {
  describe('identity', () => {
    it('Should return the same value that has been passed', () => {
      expect(_.identity(1)).toMatchSnapshot();
      expect(_.identity({})).toMatchSnapshot();
      expect(_.identity('Hackages')).toMatchSnapshot();
      expect(_.identity(1, 2, 3)).toMatchSnapshot();
    });
  });

  describe('forEachObjIndexed', () => {
    it('should map values of an object', () => {
      const result = [];
      const results2 = [];
      let name = '';
      let count = 0;
      _.forEachObjIndexed({a: 33}, (a, b) => result.push([a, b]));
      _.forEachObjIndexed({a: undefined, b: null, c: true}, (a, b) =>
        result.push([a, b])
      );
      _.forEachObjIndexed(
        {4: 42, 35: 35},
        (a, b) => (count += parseInt(a) + parseInt(b))
      );
      _.forEachObjIndexed(
        {name: 'Antonio'},
        (a, b) => (name = `My ${b} is ${a}`)
      );
      expect(result).toMatchSnapshot();
      expect(results2).toMatchSnapshot();
      expect(count).toMatchSnapshot();
      expect(name).toMatchSnapshot();
    });
  });

  describe('entries', () => {
    describe('should pair values', () => {
      it('in a function', () => {
        function Test() {
          this.a = 32;
          this.b = 45;
        }

        expect(_.entries(new Test())).toMatchSnapshot();
      });

      it('in an object', () => {
        const Test = {
          a: 32,
          b: 45
        };
        expect(_.entries(Test)).toMatchSnapshot();
      });

      it('in a class', () => {
        class Test {
          a = 32;
          b = 45;
        }
        expect(_.entries(new Test())).toMatchSnapshot();
      });
    });
  });

  describe('map', () => {
    it('it should work with the identity function', () => {
      const results = _.map([1, 2, 3, 4], x => x);
      expect(results).toMatchSnapshot();
    });
    it('it should return the index', () => {
      const results = _.map([1, 1, 1], (x, index) => index).reduce(
        (acc, next) => acc + next,
        0
      );
      expect(results).toMatchSnapshot();
    });
    it('it should return the entire array as the third parameter', () => {
      _.map([1, 2], (a, b, arr) => arr);
      expect(_.map([1, 2], (a, b, arr) => arr)[0]).toMatchSnapshot();
    });
  });

  describe('findKey', () => {
    const users = {
      barney: {age: 36, active: true},
      fred: {age: 40, active: false},
      pebbles: {age: 1, active: true}
    };

    it('should work with functions', () => {
      expect(
        _.findKey(users, function(o) {
          return o.age < 40;
        })
      ).toMatchSnapshot();
    });

    it('should work with objects', () => {
      expect(_.findKey(users, {age: 1, active: true})).toMatchSnapshot();
    });
  });

  describe('flatten', () => {
    it('should flatten', () => {
      expect(
        _.flatten([1, 2, false, {a: 32}, NaN, null, undefined])
      ).toMatchSnapshot();
      expect(
        _.flatten([[1, 2], [[false, {a: 32}], NaN], [null, [{undefined}]]])
      ).toMatchSnapshot();
      expect(_.flatten([[[[[[[[[[[[['Sorry']]]]]]]]]]]]])).toMatchSnapshot();
    });
  });

  describe('flattenDeep', () => {
    it('should flatten recursively', () => {
      expect(
        _.flattenDeep([1, 2, false, {a: 32}, NaN, null, undefined])
      ).toMatchSnapshot();
      expect(
        _.flattenDeep([[1, 2], [[false, {a: 32}], NaN], [null, [{undefined}]]])
      ).toMatchSnapshot();
      expect(
        _.flattenDeep([[[[[[[[[[[[['Sorry']]]]]]]]]]]]])
      ).toMatchSnapshot();
    });
  });

  describe('compact', () => {
    it('should compact', () => {
      expect(
        _.compact([
          1,
          2,
          false,
          {a: 32},
          NaN,
          null,
          undefined,
          false,
          null,
          0,
          '',
          undefined
        ])
      ).toMatchSnapshot();
    });
  });

  describe('chunk', () => {
    it('should chunk', () => {
      expect(
        _.chunk(['a', 'b', 12, true, false, undefined, null, 'aa'], 1)
      ).toMatchSnapshot();
      expect(
        _.chunk(['a', 'b', 12, true, false, undefined, null, 'aa'], 2)
      ).toMatchSnapshot();
      expect(
        _.chunk(['a', 'b', 12, true, false, undefined, null, 'aa'], 3)
      ).toMatchSnapshot();
      expect(_.chunk([], 2)).toMatchSnapshot();
      expect(_.chunk([42], 2)).toMatchSnapshot();
      expect(_.chunk(undefined, 4)).toMatchSnapshot();
      expect(_.chunk(32, 2)).toMatchSnapshot();
      expect(_.chunk({a: 12, b: 32, c: 45}, 2)).toMatchSnapshot();
    });
  });

  describe('zip', () => {
    it('should zip', () => {
      expect(_.zip(['a', 'b'], [1, 2], [true, false])).toMatchSnapshot();
      expect(_.zip(['a'], [1, 2], [true, true, false])).toMatchSnapshot();
      expect(_.zip(['a', 'b', 'c'], [1, 2], [true])).toMatchSnapshot();
      expect(_.zip(['hello'], [undefined])).toMatchSnapshot();
      expect(_.zip([32], 'ohoh')).toMatchSnapshot();
      expect(_.zip([32])).toMatchSnapshot();
      expect(_.zip(32)).toMatchSnapshot();
    });
  });

  describe('zipObject', () => {
    it('should zip', () => {
      expect(_.zipObject(['a', 'b'], [1, 2])).toMatchSnapshot();
      expect(_.zipObject(['a'], [1, 2])).toMatchSnapshot();
      expect(_.zipObject(['a', 'b'], [1])).toMatchSnapshot();
      expect(_.zipObject(['hello'], [true])).toMatchSnapshot();
      expect(_.zipObject([32], [false])).toMatchSnapshot();
      expect(_.zipObject([true], ['or not'])).toMatchSnapshot();
    });
  });

  describe('groupBy', () => {
    it('should group', () => {
      expect(
        _.groupBy(
          [
            {sex: 'M', name: 'Robert'},
            {sex: 'M', name: 'Bob'},
            {sex: 'F', name: 'Louise'}
          ],
          e => e.sex
        )
      ).toMatchSnapshot();
    });
  });

  describe('uniq', () => {
    it('should remove duplicate', () => {
      expect(_.uniq([1, 2, 3, 4])).toMatchSnapshot();
      expect(_.uniq([1, 1])).toMatchSnapshot();
      expect(_.uniq([null, null, true, true])).toMatchSnapshot();
      expect(_.uniq([{a: 32}, {a: 32}, {a: 32, b: 21}])).toMatchSnapshot();
    });
  });

  describe('memoize', () => {
    it('should compute the value only once', () => {
      const myMock = jest.fn((...args) => args);
      const memoized = _.memoize(myMock);
      expect(memoized(42)).toMatchSnapshot();
      expect(memoized(42)).toMatchSnapshot();
      expect(myMock).toHaveBeenCalledTimes(1);
    });

    it('should work with multiple arguments', () => {
      const myMock = jest.fn((...args) => args);
      const memoized = _.memoize(myMock);
      expect(memoized(42, 'hello', true)).toMatchSnapshot();
      expect(memoized(42, 'hello', true)).toMatchSnapshot();
      expect(myMock).toHaveBeenCalledTimes(1);
    });

    it('recompute with new arguments', () => {
      const myMock = jest.fn((...args) => args);
      const memoized = _.memoize(myMock);
      expect(memoized(42, 'hello', true)).toMatchSnapshot();
      expect(memoized(43)).toMatchSnapshot();
      expect(myMock).toHaveBeenCalledTimes(2);
    });
  });
});
