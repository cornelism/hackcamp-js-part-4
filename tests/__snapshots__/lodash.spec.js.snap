// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Lodash library chunk should chunk 1`] = `
Array [
  Array [
    "a",
  ],
  Array [
    "b",
  ],
  Array [
    12,
  ],
  Array [
    true,
  ],
  Array [
    false,
  ],
  Array [
    undefined,
  ],
  Array [
    null,
  ],
  Array [
    "aa",
  ],
]
`;

exports[`Lodash library chunk should chunk 2`] = `
Array [
  Array [
    "a",
    "b",
  ],
  Array [
    12,
    true,
  ],
  Array [
    false,
    undefined,
  ],
  Array [
    null,
    "aa",
  ],
]
`;

exports[`Lodash library chunk should chunk 3`] = `
Array [
  Array [
    "a",
    "b",
    12,
  ],
  Array [
    true,
    false,
    undefined,
  ],
  Array [
    null,
    "aa",
  ],
]
`;

exports[`Lodash library chunk should chunk 4`] = `Array []`;

exports[`Lodash library chunk should chunk 5`] = `
Array [
  Array [
    42,
  ],
]
`;

exports[`Lodash library chunk should chunk 6`] = `Array []`;

exports[`Lodash library chunk should chunk 7`] = `Array []`;

exports[`Lodash library chunk should chunk 8`] = `Array []`;

exports[`Lodash library compact should compact 1`] = `
Array [
  1,
  2,
  Object {
    "a": 32,
  },
]
`;

exports[`Lodash library entries should pair values in a class 1`] = `
Array [
  Array [
    "a",
    32,
  ],
  Array [
    "b",
    45,
  ],
]
`;

exports[`Lodash library entries should pair values in a function 1`] = `
Array [
  Array [
    "a",
    32,
  ],
  Array [
    "b",
    45,
  ],
]
`;

exports[`Lodash library entries should pair values in an object 1`] = `
Array [
  Array [
    "a",
    32,
  ],
  Array [
    "b",
    45,
  ],
]
`;

exports[`Lodash library findKey should work with functions 1`] = `"barney"`;

exports[`Lodash library findKey should work with objects 1`] = `"pebbles"`;

exports[`Lodash library flatten should flatten 1`] = `
Array [
  1,
  2,
  false,
  Object {
    "a": 32,
  },
  NaN,
  null,
  undefined,
]
`;

exports[`Lodash library flatten should flatten 2`] = `
Array [
  1,
  2,
  Array [
    false,
    Object {
      "a": 32,
    },
  ],
  NaN,
  null,
  Array [
    Object {
      "undefined": undefined,
    },
  ],
]
`;

exports[`Lodash library flatten should flatten 3`] = `
Array [
  Array [
    Array [
      Array [
        Array [
          Array [
            Array [
              Array [
                Array [
                  Array [
                    Array [
                      Array [
                        "Sorry",
                      ],
                    ],
                  ],
                ],
              ],
            ],
          ],
        ],
      ],
    ],
  ],
]
`;

exports[`Lodash library flattenDeep should flatten recursively 1`] = `
Array [
  1,
  2,
  false,
  Object {
    "a": 32,
  },
  NaN,
  null,
  undefined,
]
`;

exports[`Lodash library flattenDeep should flatten recursively 2`] = `
Array [
  1,
  2,
  false,
  Object {
    "a": 32,
  },
  NaN,
  null,
  Object {
    "undefined": undefined,
  },
]
`;

exports[`Lodash library flattenDeep should flatten recursively 3`] = `
Array [
  "Sorry",
]
`;

exports[`Lodash library forEachObjIndexed should map values of an object 1`] = `
Array [
  Array [
    33,
    "a",
  ],
  Array [
    undefined,
    "a",
  ],
  Array [
    null,
    "b",
  ],
  Array [
    true,
    "c",
  ],
]
`;

exports[`Lodash library forEachObjIndexed should map values of an object 2`] = `Array []`;

exports[`Lodash library forEachObjIndexed should map values of an object 3`] = `116`;

exports[`Lodash library forEachObjIndexed should map values of an object 4`] = `"My name is Antonio"`;

exports[`Lodash library groupBy should group 1`] = `
Object {
  "F": Array [
    Object {
      "name": "Louise",
      "sex": "F",
    },
  ],
  "M": Array [
    Object {
      "name": "Robert",
      "sex": "M",
    },
    Object {
      "name": "Bob",
      "sex": "M",
    },
  ],
}
`;

exports[`Lodash library identity Should return the same value that has been passed 1`] = `1`;

exports[`Lodash library identity Should return the same value that has been passed 2`] = `Object {}`;

exports[`Lodash library identity Should return the same value that has been passed 3`] = `"Hackages"`;

exports[`Lodash library identity Should return the same value that has been passed 4`] = `1`;

exports[`Lodash library map it should return the entire array as the third parameter 1`] = `
Array [
  1,
  2,
]
`;

exports[`Lodash library map it should return the index 1`] = `3`;

exports[`Lodash library map it should work with the identity function 1`] = `
Array [
  1,
  2,
  3,
  4,
]
`;

exports[`Lodash library memoize recompute with new arguments 1`] = `
Array [
  42,
  "hello",
  true,
]
`;

exports[`Lodash library memoize recompute with new arguments 2`] = `
Array [
  43,
]
`;

exports[`Lodash library memoize should compute the value only once 1`] = `
Array [
  42,
]
`;

exports[`Lodash library memoize should compute the value only once 2`] = `
Array [
  42,
]
`;

exports[`Lodash library memoize should work with multiple arguments 1`] = `
Array [
  42,
  "hello",
  true,
]
`;

exports[`Lodash library memoize should work with multiple arguments 2`] = `
Array [
  42,
  "hello",
  true,
]
`;

exports[`Lodash library uniq should remove duplicate 1`] = `
Array [
  1,
  2,
  3,
  4,
]
`;

exports[`Lodash library uniq should remove duplicate 2`] = `
Array [
  1,
]
`;

exports[`Lodash library uniq should remove duplicate 3`] = `
Array [
  null,
  true,
]
`;

exports[`Lodash library uniq should remove duplicate 4`] = `
Array [
  Object {
    "a": 32,
  },
  Object {
    "a": 32,
    "b": 21,
  },
]
`;

exports[`Lodash library zip should zip 1`] = `
Array [
  Array [
    "a",
    1,
    true,
  ],
  Array [
    "b",
    2,
    false,
  ],
]
`;

exports[`Lodash library zip should zip 2`] = `
Array [
  Array [
    "a",
    1,
    true,
  ],
  Array [
    undefined,
    2,
    true,
  ],
  Array [
    undefined,
    undefined,
    false,
  ],
]
`;

exports[`Lodash library zip should zip 3`] = `
Array [
  Array [
    "a",
    1,
    true,
  ],
  Array [
    "b",
    2,
    undefined,
  ],
  Array [
    "c",
    undefined,
    undefined,
  ],
]
`;

exports[`Lodash library zip should zip 4`] = `
Array [
  Array [
    "hello",
    undefined,
  ],
]
`;

exports[`Lodash library zip should zip 5`] = `
Array [
  Array [
    32,
  ],
]
`;

exports[`Lodash library zip should zip 6`] = `
Array [
  Array [
    32,
  ],
]
`;

exports[`Lodash library zip should zip 7`] = `Array []`;

exports[`Lodash library zipObject should zip 1`] = `
Object {
  "a": 1,
  "b": 2,
}
`;

exports[`Lodash library zipObject should zip 2`] = `
Object {
  "a": 1,
}
`;

exports[`Lodash library zipObject should zip 3`] = `
Object {
  "a": 1,
  "b": undefined,
}
`;

exports[`Lodash library zipObject should zip 4`] = `
Object {
  "hello": true,
}
`;

exports[`Lodash library zipObject should zip 5`] = `
Object {
  "32": false,
}
`;

exports[`Lodash library zipObject should zip 6`] = `
Object {
  "true": "or not",
}
`;
